### ISSUE TITLE
Provide access to project paths for tests

### ISSUE THREAD

**Issue by dhardy** [rust-lang/cargo#2841]
I need a test binary to read some data files from the project's repository and write some new files to a temporary location.

I've done this before with CMake, which provides variables like `CMAKE_SOURCE_DIR`, `CMAKE_CURRENT_SOURCE_DIR`, `CMAKE_BINARY_DIR` ([doc](https://cmake.org/cmake/help/v3.6/manual/cmake-variables.7.html)) which can be substituted into "configured" source code/script files or passed to tests. The distinction between source and binary directories is important especially since the source directory may not be writable; is this also the case with Cargo?

It is possible for Cargo tests to get [the path to the current binary as well as the working directory](https://users.rust-lang.org/t/how-to-reliably-locate-test-data-files-from-tests/3507), but I'd prefer a documented way of getting the project's root directory (e.g. via an environment variable or optional command-line argument), and some recommendations on where to put temporary files.

**Comment by alexcrichton** [comment]
The root directory may be accessible somehow through the `file!()` macro, and temporary files should probably go into `env!("OUT_DIR")` or ideally something managed by the test itself, Cargo doesn't really ensure that there's a temporary directory for tests. Cargo itself just shoves data into the `target/` directory.

Does that help?

**Comment by dhardy** [comment]
`file!()` just gives the path of the source relative to the project root. Doesn't help.

`OUT_DIR` doesn't appear to be set when the test is built:
```
$ cargo test --test test-suite -- --nocapture
   Compiling pippin v0.1.0 (file:///home/dhardy/other/pippin)
test-suite.rs:37:29: 37:44 error: environment variable `OUT_DIR` not defined
```
Same for most other [environment variables](http://doc.crates.io/environment-variables.html) which might contain useful paths, the exception being `CARGO_MANIFEST_DIR` which points to the "target" directory. Perhaps Cargo should define `CARGO_TARGET_DIR` and `CARGO_PROJECT_DIR` (to point to the project's top directory) when building?

I was able to get the paths I want anyway, but it's a horrible hack.

```rust
// Get absolute path to the "target" directory ("build" dir)
fn get_target_dir() -> PathBuf {
    let bin = env::current_exe().expect("exe path");
    let mut target_dir = PathBuf::from(bin.parent().expect("bin parent"));
    while target_dir.file_name() != Some(OsStr::new("target")) {
        target_dir.pop();
    }
    target_dir
}
// Get absolute path to the project's top dir, given target dir
fn get_top_dir<'a>(target_dir: &'a Path) -> &'a Path {
    target_dir.parent().expect("target parent")
}
```

**Comment by alexcrichton** [comment]
I'm somewhat wary of just keeping to pile on tons of environment variables, in general strategies like this are pretty brittle because there's no guarantee these directories exist when the binary is actually run as any number of events could have happened between compile time and execution time.

Ah but yeah I believe `OUT_DIR` is only defined when there's a build script in play, and yes `file!()` doesn't get you absolute paths.

**Comment by dhardy** [comment]
I wouldn't want to compile the output directory into a release binary, but paths are needed somehow and test binaries don't usually move. The question is which is the least brittle assumption?

1.  Do what I did in the above code: look for a parent directory named `target` from the executable's path
2.  Assume something about the directory structure inside `target` and go up a fixed number of levels from the executable's argument path
3.  Hard-code project / build directory paths (via environment variables set during compilation, or other less straightforward means)

What I'm saying is that Cargo docs should cover this somehow. I don't see anything too evil about adding a few environment variables either, so long as the documentation is kept up-to-date.

**Comment by alexcrichton** [comment]
Ah yeah of those assumptions I think (2) is the least brittle, and I'd be totally fine documenting it!

**Comment by dhardy** [comment]
Since [Cargo supports out-of-source builds](https://internals.rust-lang.org/t/im-moving-away-heres-my-two-cents/3767/4?u=dhardy) neither options (1) or (2) will work reliably.

IMO "trying not to pile on a ton of environment variables" sounds a bit like Cargo is trying to bind the hands of developers, rather than help them. Cargo is just a tool and should include the features users need.

**Comment by xftroxgpx** [comment]
**tl;dr:** would be nice to have:
- `CARGO_WORKSPACE_DIR` - which could be a parent dir of `CARGO_MANIFEST_DIR` for example.
- `CARGO_TARGET_DIR` - fullpath pointer to `target/` or to `target/debug/` ? unsure (assuming former)
- `CARGO_PKG_OUTFILE` - output exe filename full path (or just fname, if `CARGO_TARGET_DIR` is also available)

**long read:**
From what I can tell, `CARGO_MANIFEST_DIR` and `CARGO_PROJECT_DIR` are the same. 
It's `CARGO_TARGET_DIR` which could differ significantly, under a workspace.
For example:
if `CARGO_MANIFEST_DIR`=`CARGO_PROJECT_DIR`=`/home/xftroxgpx/build/2nonpkgs/rust.stuff/rustlearnage/recompile_self`
then you would think that `CARGO_TARGET_DIR` would be equal to `$CARGO_MANIFEST_DIR/target/$PROFILE`  (assuming you look at it from within a running `build.rs`, else you don't have `$PROFILE` [it seems](http://doc.crates.io/environment-variables.html#environment-variables-cargo-sets-for-build-scripts))
which resolves to `/home/xftroxgpx/build/2nonpkgs/rust.stuff/rustlearnage/recompile_self/target/debug`.
However, since this dir `/home/xftroxgpx/build/2nonpkgs/rust.stuff/rustlearnage/` is a workspace dir (with its own Cargo.tml file, example [here](https://github.com/xftroxgpx/rustlearnage/blob/d6ce52e8a03ad1b33cb5d360c52c1becce128320/Cargo.toml#L1)), then in actuality
`CARGO_TARGET_DIR` has the value `$CARGO_WORKSPACE_DIR/target/$PROFILE` aka `/home/xftroxgpx/build/2nonpkgs/rust.stuff/rustlearnage/target/debug/`  and NOT `/home/xftroxgpx/build/2nonpkgs/rust.stuff/rustlearnage/recompile_self/target/debug`
(of course `CARGO_WORKSPACE_DIR` does NOT exist currently as an env. var (ie. not implemented, just like `CARGO_TARGET_DIR`) )

If `CARGO_TARGET_DIR` would exist, then I could hardcode(in the exe) the exe file name like: `$CARGO_TARGET_DIR/target/$PROFILE/$CARGO_PKG_NAME` (though it feels hacky!), unless there's some way to specify the exe filename in Cargo.toml (which I'm sure there is, but I forgot how, currently) in which case I would kindly request a way to get that one too, at compile time. Or perhaps even better, a `CARGO_PKG_OUTFILE` = fullpath to exe, at compile time.


Basically `CARGO_PKG_OUTFILE` would be the same value as the resulting binary filename that `cargo run` executes.
Any pointers on where to look in the source code for implemeting this? Even if a PR won't be accepted.

As to why would I need this for my particular case: I need to hardcode the compiledexe fullpath filename  so that when I detect hardlinks to this exe(by other means), I can then show a message pointing to the real exe which would be the only updated(recompiled) one. (the hardlink binary would be the outdated one)
This means `std::env::current_exe` will point to the same name as the hardlink (in Linux) which is why I need the hardcoded fullpath of the executable at compile time.

**Comment by dhardy** [comment]
This issue is definitely stale (no active interest), but I don't believe it was solved. There are a [host of environment variables](https://doc.rust-lang.org/cargo/reference/environment-variables.html) available, but I don't see a documented way of accessing the source path or creating temporary files for tests.

[Build scripts](https://doc.rust-lang.org/cargo/reference/build-scripts.html) do have documented access to the source (the working directory) and the output (`OUT_DIR`), but there is no indication that this applies to test resources.

And now I've just spend five minutes summarizing a stale issue because of this bot; not sure whether that's a good or bad thing.

**Comment by dwijnand** [comment]
I've found summaries help move issues forward, so thank you!

**Comment by khoek** [comment]
Because this was quite a while ago now: has anything changed since this issue was created?

Since Cargo has tests for itself and (presumably) modifies the filesystem during them, how does Cargo handle it?

**Comment by ehuss** [comment]
Reading files from the project can be done by accessing `env!("CARGO_MANIFEST_DIR")`. 

For writing, using a temp directory is probably the safest choice. The [`tempfile`](https://crates.io/crates/tempfile) crate can be useful for that.

Cargo's testsuite writes to the `target` directory by looking at the location of the test executable (this is an approach mentioned above). That is not necessarily the safest thing to do, and is not explicitly documented.

**Comment by rubik** [comment]
Has there been any progress on this? I've just hit the same issue: my tests need external data files and I need to access the path to those files from the tests.

**Comment by vojtechkral** [comment]
Now that #9375 is merged, I'd propose to close this issue, seems like it's been addressed.

- To access your source files, you can use the `CARGO_MANIFEST_DIR` env var
- To access a temporary location suitable for writing test files, use `CARGO_TARGET_TMPDIR` in integration tests (and benches). This is a compile time env var only.
  `CARGO_TARGET_TMPDIR` is also brand new, so it will take some weeks before it hits stable toolchains.
  In the meantime, you can use `option_env!("CARGO_TARGET_TMPDIR")` to detect if it's available. If not, as a fallback you can use something like `env::current_exe()/../../tmp` or `$CARGO_MANIFEST_DIR/target/tmp` or similar.

**Comment by dhardy** [comment]
Thanks for the summary @vojtechkral. I'll close as recommended since I currently have no use-case to test with.
